tests/phpunit/CRM/Civicase/Service/CaseCategoryInstanceTest.php

```diff
@@ -0,0 +1,117 @@
+<?php
+
+use CRM_Civicase_Service_CaseCategoryInstance as CaseCategoryInstanceService;
+use CRM_Civicase_Test_Fabricator_CaseCategoryInstance as CaseCategoryInstanceFabricator;
+use CRM_Civicase_Test_Fabricator_CaseCategory as CaseCategoryFabricator;
+
+/**
+ * Runs tests on CaseCategoryInstance Service tests.
+ *
+ * @group headless
+ */
+class CRM_Civicase_Service_CaseCategoryInstanceTest extends BaseHeadlessTest {
+
+  /**
+   * Test that the category instances list is correctly formatted.
+   */
+  public function testGetCategoryInstances() {
+    // Generate random category instances.
+    $categoryInstanceOne = CaseCategoryInstanceFabricator::fabricate();
+    $categoryInstanceTwo = CaseCategoryInstanceFabricator::fabricate();
+    $caseCategoryInstanceService = new CaseCategoryInstanceService();
+
+    // We call the method for getting the list.
+    $caseCategoryInstances = $caseCategoryInstanceService->getCaseCategoryInstances();
+
+    // And we verify the structure and content of the list received.
+    $this->assertArrayHasKey($categoryInstanceOne['id'], $caseCategoryInstances);
+    $this->assertArrayHasKey($categoryInstanceTwo['id'], $caseCategoryInstances);
+    $this->assertEquals(
+      $categoryInstanceOne['category_id'],
+      $caseCategoryInstances[$categoryInstanceOne['id']]->category_id
+    );
+    $this->assertEquals(
+      $categoryInstanceOne['instance_id'],
+      $caseCategoryInstances[$categoryInstanceOne['id']]->instance_id
+    );
+    $this->assertEquals(
+      $categoryInstanceTwo['category_id'],
+      $caseCategoryInstances[$categoryInstanceTwo['id']]->category_id
+    );
+    $this->assertEquals(
+      $categoryInstanceTwo['instance_id'],
+      $caseCategoryInstances[$categoryInstanceTwo['id']]->instance_id
+    );
+  }
+
+  /**
+   * Test that the category instances list by name is correctly formatted.
+   */
+  public function testGetCategoryInstancesByInstanceName() {
+    // Generate random category instances.
+    CaseCategoryInstanceFabricator::fabricate();
+    CaseCategoryInstanceFabricator::fabricate();
+    // And get a random instance type.
+    $categoryInstanceType = $this->getRandomCategoryInstanceType();
```

Comment:
I think this method should not return a random case category instance but should rather accept the instance type value i.e (rename getRandomCategoryInstanceType to something better)
```php
 $categoryInstanceOne =  CaseCategoryInstanceFabricator::fabricate();
  $categoryInstanceTwo  = CaseCategoryInstanceFabricator::fabricate();
$caseCategoryInstanceType = $this->getRandomCategoryInstanceType($categoryInstanceOne['instance_id');
```

This way when the $categoryInstanceType['name'] is passed into the function, we can verify that other properties like the instance_id and category_id matches i.e

```php
//Check that only one record is returned when searching by the name.
$this->assertEquals(1,  $caseCategoryInstances);

$this->assertEquals($categoryInstanceOne['instance_id'], $categoryInstances[0]->instance_id);
$this->assertEquals($categoryInstanceOne['category_id'], $categoryInstance[0]->category_id);
```

let me know what you think.

Reply:
@tunbola I agree with that, and it is the correct way to check if a method that filters something is correctly working. I was undecided on creating or not more instance types, because I discarded the idea of the fabricator, but now I have added a method for generating these instances on the test class itself.
I added comments on the code, but let me know if something is not clear enough.